google/cloud/aiplatform/featurestore/featurestore.py

```diff
@@ -557,3 +562,597 @@ def create_entity_type(
+                tfrecord_destination=destination
+            )
+
+        if isinstance(read_instances, gca_io.BigQuerySource):
+            batch_read_feature_values_request.bigquery_read_instances = read_instances
+        elif isinstance(read_instances, gca_io.CsvSource):
+            batch_read_feature_values_request.csv_read_instances = read_instances
+
+        if pass_through_fields is not None:
+            batch_read_feature_values_request.pass_through_fields = [
+                gca_featurestore_service.BatchReadFeatureValuesRequest.PassThroughField(
+                    field_name=pass_through_field
+                )
+                for pass_through_field in pass_through_fields
+            ]
+
+        return batch_read_feature_values_request
+
+    def _get_read_instances(
+        self, read_instances: Union[str, List[str]],
+    ) -> Union[gca_io.BigQuerySource, gca_io.CsvSource]:
+        """Gets read_instances
+
+        Args:
+            read_instances (Union[str, List[str]]):
+                Required. Read_instances can be either BigQuery URI to the input table,
+                or Google Cloud Storage URI(-s) to the csv file(s).
+
+        Returns:
+            Union[gca_io.BigQuerySource, gca_io.CsvSource] - BigQuery source or Csv source for read instances.
+
+        Raises:
+            TypeError if read_instances is not a string or a list of strings.
+            ValueError if read_instances uri does not start with 'bq://' or 'gs://'.
+            ValueError if uris in read_instances do not start with 'gs://'.
+        """
+        if isinstance(read_instances, str):
+            if not (
+                read_instances.startswith("bq://") or read_instances.startswith("gs://")
+            ):
+                raise ValueError(
+                    "The read_instances accepts a single uri starts with 'bq://' or 'gs://'."
+                )
+        elif isinstance(read_instances, list) and all(
+            [isinstance(e, str) for e in read_instances]
+        ):
+            if not all([e.startswith("gs://") for e in read_instances]):
+                raise ValueError(
+                    "The read_instances accepts a list of uris start with 'gs://' only."
+                )
+        else:
+            raise TypeError(
+                "The read_instances type should to be either a str or a List[str]."
+            )
+
+        if isinstance(read_instances, str):
+            if read_instances.startswith("bq://"):
+                return gca_io.BigQuerySource(input_uri=read_instances)
+            else:
+                read_instances = [read_instances]
+
+        return gca_io.CsvSource(gcs_source=gca_io.GcsSource(uris=read_instances))
+
+    @base.optional_sync(return_input_arg="self")
+    def batch_serve_to_bq(
+        self,
+        bq_destination_output_uri: str,
+        entity_type_ids: List[str],
+        entity_type_destination_fields: Optional[
```

Comment:
This argument name is a bit confusing. It is to specify what features to read, and (optional) the output column name, right? May be change it to smt related to feature than entity type, e.g `feature_to_read` or some sort.

Alternatively, we can have a separate struct to define the entity_type and what feature to read within that entity type, and provide a list of that to this function, instead of the separate entity_type_ids and this entity_type_destination_field.


Another option is, we can have a flat list of feature to read, using the full-resource path name. Internally, we can group them by entity type and construct the read request. This could be a little bit repetitive on the calling side, but I think the interface is more intuitive. We'll need to use different struct for 'read_instance' though (e.g Dict instead of List)

Reply:
Isn't `read_instances` the bq uri or gcs uri(s)? How will dict work here? 

Regarding renaming the `entity_type_destination_fields`, how about `feature_destination_fields` to match the `feature_source_fields` for ingestion in EntityType?

Regarding the data type and usage of the `entity_type_destination_fields`:
1) I think the current implementation is trying to implement the first alternative suggested here. The purpose for having a separate `entity_type_ids` and `entity_type_destination_fields` is that for some `entity_type` that user want to batch serve all features in it, without a different output feature name, user won't need to include a list of features or a mapping (feature to output name) for that entity_type in the `entity_type_destination_fields`. We can remove the separated `entity_type_ids` field and make it required to have all `entity_type_id` specified in the `entity_type_destination_fields`'s key, with ["*"] value if all features are to be served. For example,
```
{
    "entity_type_1": ["*"], 
    "entity_type_2": {"feature_1": "feature_1_col"}, 
    "entity_type_3": {"feature_1", "feature_2"}, 
}
```

2) the second option is a bit redundant but it is indeed very clear and straight forward
{"...path/to/fs/fs1/et/e1/f/f1": "f11_col", "...path/to/fs/fs1/et/e2/f/f1": "f21_col"}

Will the modification of the current implementation in 1) be less confusing and more appealing for usage?

Reply:
Regarding [1], you also have the option to support mapping the output column name to something different than feature name. I feel this overloaded argument is hard for users to understand (though I may agree it could be convenient once user get a good grasp of it)

I think it is easier to understand to split into 2 arguments: one to define the feature to read, the other one to define the  (overridden) output column name. This also allow user to define the output column only for the feature they want to (i.e they want to read */all features from entity_type_1 and output to the column with the same name, except feature `foo` will be output to column `bar`) 

With the flat-list, the entity type IDs can be extracted from the feature path name. However, unless we put some constraints that the list have to group by entity type, the order of the entity type in the `read_instance` could be ambiguous. Thus I think using Dict (e.g `{'entity_type_1' : 'user_1234', 'entity_type_2': 'product_xyz' }` is clearer than using array (`['user_1234', 'product_xyz']`). The flat list option also requires all to be explicitly listed, i.e "*" is not supported. This is the biggest cons for that option.

Reply:
removed ```entity_type_ids``` and ```entity_type_destination_fields``` to avoid confusion, added two dict ```serving_feature_ids``` for an entity_type / features mapping and ```feature_destination_fields ``` for feature_id / feature_destination_field_name mapping.